diff --git a/modules/libpref/Preferences.cpp b/modules/libpref/Preferences.cpp
index c04744a..f030c23 100644
--- a/modules/libpref/Preferences.cpp
+++ b/modules/libpref/Preferences.cpp
@@ -3598,6 +3598,7 @@ static bool TelemetryPrefValue() {
   // For platforms with Unified Telemetry (here meaning not-Android),
   // toolkit.telemetry.enabled determines whether we send "extended" data.
   // We only want extended data from pre-release channels due to size.
+  return false;
 
   constexpr auto channel = MOZ_STRINGIFY(MOZ_UPDATE_CHANNEL) ""_ns;
 
